bridgenode/rev.go

```diff
@@ -62,19 +64,166 @@ func BlockAndRevReader(
+	if err != nil {
+		return
+	}
+	defer offsetFile.Close() // file always closes
+
+	// offset file consists of 12 bytes per block
+	// tipnum * 12 gives us the correct position for that block
+	_, err = offsetFile.Seek(int64(12*startAt), 0)
+	if err != nil {
+		return
 	}
+	offsetReader := bufio.NewReaderSize(offsetFile, int(12*count))
+
+	offsets := make([]uint32, count)
+	revOffsets := make([]uint32, count)
+
+	var datFileNum uint32
+	// minOffset holds the smallest block offset that was seen
+	minOffset := uint32(math.MaxUint32)
+	// maxOffset holds the largest block offset that was seen
+	maxOffset := uint32(0)
+	// minRevOffset holds the smallest rev offset that was seen
+	minRevOffset := uint32(math.MaxUint32)
+	// maxRevOffset holds the largest rev offset that was seen
+	maxRevOffset := uint32(0)
+	// offsetsRead holds the number of blocks < count that are loacated in
+	// blk file with number `datFileNum`
+	offsetsRead := uint32(0)
+	for ; offsetsRead < uint32(count); offsetsRead++ {
+		var datFileNumTmp uint32
+		err = binary.Read(offsetReader, binary.BigEndian, &datFileNumTmp)
+		if err != nil {
+			break
+		}
+		if offsetsRead > 0 && datFileNumTmp != datFileNum {
+			// break if the block is located in a different blk file
+			break
+		}
+		datFileNum = datFileNumTmp
+		err = binary.Read(offsetReader, binary.BigEndian, &offsets[offsetsRead])
+		if err != nil {
+			return
+		}
+		err = binary.Read(offsetReader, binary.BigEndian, &revOffsets[offsetsRead])
+		if err != nil {
+			return
+		}
+
+		if offsets[offsetsRead] < minOffset {
+			minOffset = offsets[offsetsRead]
+		}
+		if offsets[offsetsRead] > maxOffset {
+			maxOffset = offsets[offsetsRead]
+		}
+
+		if revOffsets[offsetsRead] < minRevOffset {
+			minRevOffset = revOffsets[offsetsRead]
+		}
+		if revOffsets[offsetsRead] > maxRevOffset {
+			maxRevOffset = revOffsets[offsetsRead]
+		}
+	}
+
+	if offsetsRead == 0 {
+		return
+	}
+
+	blockFile, err := os.Open(filepath.Join(blockDir,
+		fmt.Sprintf("blk%05d.dat", datFileNum)))
+	if err != nil {
+		return
+	}
+	defer blockFile.Close()
+	blockFileInfo, _ := blockFile.Stat()
+
+	// Read all block data needed for the blocks into memory.
+	// The blocks that are needed are located between the minOffset and
+	// maxOffset+(the size of the block at maxOffset).
+	// 1<<23 = +8MB TODO: whats the max on disk size of a block?
+	blockData := make([]byte, min(maxOffset+1<<23, uint32(blockFileInfo.Size()))-minOffset)
+	_, err = blockFile.ReadAt(blockData, int64(minOffset))
+	if err != nil {
+		return
+	}
+
+	revFile, err := os.Open(filepath.Join(blockDir,
+		fmt.Sprintf("rev%05d.dat", datFileNum)))
+	if err != nil {
+		return
+	}
+	defer revFile.Close()
+	revFileInfo, _ := revFile.Stat()
+
+	// Read all rev data needed for the blocks into memory.
+	// The rev data thats needed is located between the minRevOffset and
+	// maxRevOffset+(the size of the rev at maxRevOffset).
+	// 1<<23 = +8MB TODO: whats the max on disk size of a rev?
```

Comment:
Rev blocks contain the spend TXOs so the limit is on them. The theoretical limit is 240MB as a Bitcoin block can have a max of 23487 inputs with theoretical max of 10KB per UTXO.

But realistically no one creates UTXOs that big so 128MB here would suffice as well. I don't think there are rev files that are bigger than ~50MB so I think we're safe.